processing/src/main/java/org/apache/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -275,6 +367,63 @@ Sequence<ScanResultValue> stableLimitingSort(
     }
   }
 
+  Sequence<ScanResultValue> multiColumnSort(
+      Sequence<ScanResultValue> inputSequence,
+      ScanQuery scanQuery
+  ) throws IOException
+  {
+
+    if (scanQuery.getScanRowsLimit() > Integer.MAX_VALUE) {
+      throw new UOE(
+          "Limit of %,d rows not supported for priority queue strategy of non-time-ordering scan results",
+          scanQuery.getScanRowsLimit()
+      );
+    }
+    // Converting the limit from long to int could theoretically throw an ArithmeticException but this branch
+    // only runs if limit < MAX_LIMIT_FOR_IN_MEMORY_TIME_ORDERING (which should be < Integer.MAX_VALUE)
+    int limit = Math.toIntExact(scanQuery.getScanRowsLimit());
+    List<String> sortColumns = scanQuery.getOrderBys().stream().map(orderBy -> orderBy.getColumnName()).collect(Collectors.toList());
+    MultiColumnSorter<ScanResultValue> multiColumnSorter = new QueueBasedMultiColumnSorter<>(limit, scanQuery.getOrderByNoneTimeResultOrdering());
+    Yielder<ScanResultValue> yielder = Yielders.each(inputSequence);
+    try {
+      boolean doneScanning = yielder.isDone();
+      // We need to scan limit elements and anything else in the last segment
+      while (!doneScanning) {
+        ScanResultValue next = yielder.get();
+        List<ScanResultValue> singleEventScanResultValues = next.toSingleEventScanResultValues();
+        for (ScanResultValue srv : singleEventScanResultValues) {
+          // Using an intermediate unbatched ScanResultValue is not that great memory-wise, but the column list
+          // needs to be preserved for queries using the compactedList result format
+          List<Integer> idxs = sortColumns.stream().map(c -> srv.getColumns().indexOf(c)).collect(Collectors.toList());
+          List events = (List) (srv.getEvents());
+          for (Object event : events) {
+            List<Comparable> sortValues;
+            if (event instanceof LinkedHashMap) {
+              sortValues = sortColumns.stream().map(c -> ((LinkedHashMap<Object, Comparable>) event).get(c)).collect(Collectors.toList());
+            } else {
+              sortValues = idxs.stream()
+                               .map(idx -> ((List<Comparable>) event).get(idx))
+                               .collect(Collectors.toList());
+            }
+            multiColumnSorter.add(new MultiColumnSorter.MultiColumnSorterElement<>(srv, sortValues));
+          }
+        }
+        yielder = yielder.next(null);
+        doneScanning = yielder.isDone();
+      }
+      final List<ScanResultValue> sortedElements = new ArrayList<>(multiColumnSorter.size());
+      Iterators.addAll(sortedElements, multiColumnSorter.drainElement());
+      return Sequences.simple(sortedElements);
```

Comment:
Given the items above, and the code immediately above, it looks like what this is doing is:

1. Materialize rows from the cursor into an array.
2. Create an indirection vector, and sort the list via that indirection.
3. Populate an array with the sorted rows.
4. Limit the number of rows returned from the sorted list.

Given this, we can probably make the code a bit simpler.